### ISSUE TITLE
Length of named tuple not correctly obtained as self

### ISSUE THREAD

**Issue by Vegemash** [microsoft/pyright#2317]
When subclassing the typing.NamedTuple and using the star syntax for function calls pyright incorrectly throws an error about the number of arguments.

**Describe the bug**

**To Reproduce**
Steps to reproduce the behavior.
Run pyright on the below code

**Expected behavior**
There should be no error as the length of AB is known

**Screenshots or Code**
``` python
from __future__ import annotations

from typing import NamedTuple


class ABC(NamedTuple):
    a: float
    b: float
    c: float

    def to_rgba(self) -> ABC:
        return ABC(*self)


class AB(NamedTuple):
    a: float
    b: float

    def to_abc(self) -> ABC:
        return ABC(*self, 1)  # "Expected 3 positional arguements"

    def test(self) -> ABC:
        tup = (1, 2)
        return ABC(*tup, 1)  # Does not produce the error

    def test2(self) -> ABC:
        tup = AB(1, 2)
        return ABC(*tup, 1)  # Does not produce the error
```

**VS Code extension or command-line**
Command line

**Additional context**
Love the work guys :)

**Comment by erictraut** [comment]
Thanks for the bug report. The reason this wasn't working correctly is that `self` had an implied generic type. If the `self` parameter isn't explicitly annotated, we give it the type `T = TypeVar("T", bound=AB)` (i.e. a generic type that is bound to the containing class). The logic for unpacking arguments wasn't working with generic types, so `*self` wasn't correctly interpreted as a tuple with a known length. I've addressed the bug, and the fix will be in the next release.

For what it's worth, I'd recommend against using `*self` in this context. It's fragile because the class `AB` is not marked `@final`, which means that someone could come along and subclass it, adding more named tuple entries. Calling `to_abc` would then crash at runtime, and a static type checker wouldn't detect the problem. It would be safer to use `ABC(self.a, self.b, 1)`.

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.170, which I just published. It will also be included in the next release of pylance.
